### ISSUE TITLE
Touchover / Touchout possible?

### ISSUE THREAD

**Issue by SebastianNette** [GoodBoyDigital/pixi.js#616]
Hi guys,

I thought that it would be pretty nice to have touchover and touchout events for pixi.
The code could be the same as the code for mouseover and mouseout so I'm pretty confident that you can implement that.

Also it would be pretty neat if the tap happens on the currently hovered element at touchend. (Maybe with a setting to disable tap at touchend if someone doesn't like it)

It sure would feel more lively and it makes hitting buttons much easier if the gui is small and you dont have a lot space to put buttons far away from eachother.

So in short: Call touchstart of a sprite/object if the element is hovered via touchmove (stage) and tap on touchend.

**Comment by GoodBoyDigital** [comment]
Keeping to the way the DOM interaction modal works, we will not be adding touch / over / out i'm afraid. But as you say it should be pretty easy to implement if required. Thanks! :+1:

**Comment by weepy** [comment]
heya - we need to add this. You mentioned it would be "easy to implement if required." . Just wondering if you have any pointers of where to start ?

**Comment by alanthai** [comment]
According to the [Touch API](https://developer.mozilla.org/en-US/docs/Web/API/Touch_events#Setting_up_the_event_handlers) there's no such thing as `touchout` or` touchover` events.

I still needed that type of interaction so here's a monkey patch I wrote that works for me. It works the same way as `mouseout` and `mouseover`.

```
function isMobile() {
    var check = false;
    (function(a){if(/(android|bb\d+|meego).+mobile|avantgo|bada\/|blackberry|blazer|compal|elaine|fennec|hiptop|iemobile|ip(hone|od)|iris|kindle|lge |maemo|midp|mmp|mobile.+firefox|netfront|opera m(ob|in)i|palm( os)?|phone|p(ixi|re)\/|plucker|pocket|psp|series(4|6)0|symbian|treo|up\.(browser|link)|vodafone|wap|windows ce|xda|xiino/i.test(a)||/1207|6310|6590|3gso|4thp|50[1-6]i|770s|802s|a wa|abac|ac(er|oo|s\-)|ai(ko|rn)|al(av|ca|co)|amoi|an(ex|ny|yw)|aptu|ar(ch|go)|as(te|us)|attw|au(di|\-m|r |s )|avan|be(ck|ll|nq)|bi(lb|rd)|bl(ac|az)|br(e|v)w|bumb|bw\-(n|u)|c55\/|capi|ccwa|cdm\-|cell|chtm|cldc|cmd\-|co(mp|nd)|craw|da(it|ll|ng)|dbte|dc\-s|devi|dica|dmob|do(c|p)o|ds(12|\-d)|el(49|ai)|em(l2|ul)|er(ic|k0)|esl8|ez([4-7]0|os|wa|ze)|fetc|fly(\-|_)|g1 u|g560|gene|gf\-5|g\-mo|go(\.w|od)|gr(ad|un)|haie|hcit|hd\-(m|p|t)|hei\-|hi(pt|ta)|hp( i|ip)|hs\-c|ht(c(\-| |_|a|g|p|s|t)|tp)|hu(aw|tc)|i\-(20|go|ma)|i230|iac( |\-|\/)|ibro|idea|ig01|ikom|im1k|inno|ipaq|iris|ja(t|v)a|jbro|jemu|jigs|kddi|keji|kgt( |\/)|klon|kpt |kwc\-|kyo(c|k)|le(no|xi)|lg( g|\/(k|l|u)|50|54|\-[a-w])|libw|lynx|m1\-w|m3ga|m50\/|ma(te|ui|xo)|mc(01|21|ca)|m\-cr|me(rc|ri)|mi(o8|oa|ts)|mmef|mo(01|02|bi|de|do|t(\-| |o|v)|zz)|mt(50|p1|v )|mwbp|mywa|n10[0-2]|n20[2-3]|n30(0|2)|n50(0|2|5)|n7(0(0|1)|10)|ne((c|m)\-|on|tf|wf|wg|wt)|nok(6|i)|nzph|o2im|op(ti|wv)|oran|owg1|p800|pan(a|d|t)|pdxg|pg(13|\-([1-8]|c))|phil|pire|pl(ay|uc)|pn\-2|po(ck|rt|se)|prox|psio|pt\-g|qa\-a|qc(07|12|21|32|60|\-[2-7]|i\-)|qtek|r380|r600|raks|rim9|ro(ve|zo)|s55\/|sa(ge|ma|mm|ms|ny|va)|sc(01|h\-|oo|p\-)|sdk\/|se(c(\-|0|1)|47|mc|nd|ri)|sgh\-|shar|sie(\-|m)|sk\-0|sl(45|id)|sm(al|ar|b3|it|t5)|so(ft|ny)|sp(01|h\-|v\-|v )|sy(01|mb)|t2(18|50)|t6(00|10|18)|ta(gt|lk)|tcl\-|tdg\-|tel(i|m)|tim\-|t\-mo|to(pl|sh)|ts(70|m\-|m3|m5)|tx\-9|up(\.b|g1|si)|utst|v400|v750|veri|vi(rg|te)|vk(40|5[0-3]|\-v)|vm40|voda|vulc|vx(52|53|60|61|70|80|81|83|85|98)|w3c(\-| )|webc|whit|wi(g |nc|nw)|wmlb|wonu|x700|yas\-|your|zeto|zte\-/i.test(a.substr(0,4)))check = true})(navigator.userAgent||navigator.vendor||window.opera);
    return check;
}

function patchInteractionManager ()
{
  var proto = PIXI.interaction.InteractionManager.prototype;

  function extendMethod (method, extFn) {
      var old = proto[method];
      proto[method] = function () {
          old.call(this, ...arguments);
          extFn.call(this, ...arguments);
      };
  }

  extendMethod('onTouchMove', function () {
      this.didMove = true;
  });

  if (isMobile()) {
      proto.update = mobileUpdate;
  }

  function mobileUpdate(deltaTime) {
    this._deltaTime += deltaTime;

    if (this._deltaTime < this.interactionFrequency)
    {
        return;
    }

    this._deltaTime = 0;

    if (!this.interactionDOMElement)
    {
        return;
    }

    // if the user move the mouse this check has already been done using the mouse move!
    if(this.didMove)
    {
        this.didMove = false;
        return;
    }

    this.processInteractive(this.eventData.data.global, this.renderer._lastObjectRendered, this.processTouchOverOut, true );
  }

  // patch processTouchMove
  extendMethod('processTouchMove', function(displayObject, hit) {
      this.processTouchOverOut(displayObject, hit);
  });

  // create processTouchOverOut
  proto.processTouchOverOut = function (displayObject, hit)
  {
    if(hit)
    {
        if(!displayObject._over)
        {
            displayObject._over = true;
            this.dispatchEvent( displayObject, 'touchover', this.eventData );
        }

    }
    else
    {
        if(displayObject._over)
        {
            displayObject._over = false;
            this.dispatchEvent( displayObject, 'touchout', this.eventData);
        }
    }
  };
}
patchInteractionManager();
```
